test/reply-trailers.test.js

```diff
@@ -110,18 +112,135 @@ test('send trailers when payload is json', t => {
+
+    fastify.get('/', function (request, reply) {
+      reply.trailer('ETag', function (reply, payload, callback) {
+        let value = ''
+        t.same(typeof payload.pipe === 'function', true)
+        payload.on('data', (chunk, encoding) => {
+          value += chunk.toString(encoding)
+        })
+        payload.on('end', () => {
+          t.same(value, JSON.stringify({ hello: 'world' }))
+          callback(null, 'custom-etag')
+        })
+      })
+      const stream = Readable.from([JSON.stringify({ hello: 'world' })])
+      reply.send(stream)
+    })
+
+    fastify.inject({
+      method: 'GET',
+      url: '/'
+    }, (error, res) => {
+      t.error(error)
+      t.equal(res.statusCode, 200)
+      t.equal(res.headers['transfer-encoding'], 'chunked')
+      t.equal(res.headers.trailer, 'etag')
+      t.equal(res.trailers.etag, 'custom-etag')
+      t.notHas(res.headers, 'content-length')
+    })
+  })
+
+  test('multiple trailers', t => {
+    t.plan(11)
+
+    const fastify = Fastify()
+
+    fastify.get('/', function (request, reply) {
+      reply.trailer('ETag', function (reply, payload, callback) {
+        let value = ''
+        t.same(typeof payload.pipe === 'function', true)
+        payload.on('data', (chunk, encoding) => {
+          value += chunk.toString(encoding)
+        })
+        payload.on('end', () => {
+          t.same(value, JSON.stringify({ hello: 'world' }))
+          callback(null, 'custom-etag')
+        })
+      })
+      reply.trailer('Foo', function (reply, payload, callback) {
+        let value = ''
+        t.same(typeof payload.pipe === 'function', true)
+        payload.on('data', (chunk, encoding) => {
+          value += chunk.toString(encoding)
+        })
+        payload.on('end', () => {
+          t.same(value, JSON.stringify({ hello: 'world' }))
+          callback(null, 'bar')
+        })
+      })
+      const stream = Readable.from([JSON.stringify({ hello: 'world' })])
+      reply.send(stream)
+    })
+
+    fastify.inject({
+      method: 'GET',
+      url: '/'
+    }, (error, res) => {
+      t.error(error)
+      t.equal(res.statusCode, 200)
+      t.equal(res.headers['transfer-encoding'], 'chunked')
+      t.equal(res.headers.trailer, 'etag foo')
+      t.equal(res.trailers.etag, 'custom-etag')
+      t.equal(res.trailers.foo, 'bar')
+      t.notHas(res.headers, 'content-length')
+    })
+  })
+})
+
+test('async-await trailers', (t) => {
   t.plan(7)
 
   const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  const data = JSON.stringify({ hello: 'world' })
+  const hash = createHash('md5')
+  hash.update(data)
+  const md5 = hash.digest('hex')
 
   fastify.get('/', function (request, reply) {
-    reply.trailer('ETag', function (reply, payload) {
-      t.same(payload, null)
-      return 'custom-etag'
+    reply.trailer('Content-MD5', async function (reply, payload) {
+      t.equal(data, payload)
+      await sleep(500)
+      return md5
     })
```

Comment:
The `async-await` seems to be buggy here.
I believe it is not the problem of `node` itself since a plain `http` test is passed.

```js
import { createHash } from 'crypto'
import http from 'http'
import net from 'net'
import { Readable } from 'stream'
import { setTimeout as setTimeoutPromise } from 'timers/promises'

const server = http.createServer(function(request, response) {
  response.writeHead(200, {
    'Content-Type': 'application/json; charset=utf-8',
    'Transfer-Encoding': 'chunked',
    Trailer: "Content-MD5"
  })
  const stream = Readable.from([JSON.stringify({ hello: 'world' })])
  stream.pipe(response, { end: false })
  
  setTimeout(async () => {
    const hasher = createHash('MD5')
    hasher.update(JSON.stringify({ hello: 'world' }))
    await setTimeoutPromise(2000)
    response.addTrailers({
      "Content-MD5": hasher.digest().toString('hex')
    })
    response.end()
  }, 1000)
})
server.listen(function() {
  const client = net.createConnection({ port: server.address().port })
  client.write('GET / HTTP/1.1\r\n\r\n')
  client.setEncoding('utf8')
  client.on('data', data => {
    console.log(data)
  })
  client.on('end', () => {
    console.log('end')
    client.end()
    server.close()
  })
})
```

I think someone can point to me why `addTrailers` doesn't sent in this case.

Reply:
Is this test failing? What's the problem?

Reply:
This is the only failing test, I cannot figure out yet. Given that https://github.com/fastify/fastify/pull/4380 is passed without problem.

One possible though is my usage of [cloneable-readable](https://github.com/mcollina/cloneable-readable) is wrong.

Reply:
The implementation is the same in two repo.
Only difference is stream operations.

Need to dig in further.